### PR TITLE
Implement ANY aggregator

### PR DESCRIPTION
**Pull request by maytasm3**
Implement ANY aggregator

### Description

Implement ANY function that returns the first value that we can find

- AnySqlAggregator can reuse a lot of code in EarliestLatestSqlAggregator. Hence, added ANY to EarliestLatestSqlAggregator and rename EarliestLatestSqlAggregator to EarliestLatestAnySqlAggregator. (Maybe there can be a better name for that)

- SQL function keyword is ANY_VALUE since ANY is SQL reserved already.

- Create LongAnyAggregatorFactory, FloatAnyAggregatorFactory, DoubleAnyAggregatorFactory, StringAnyAggregatorFactory. The combine logic will just return lhs if not null otherwise return rhs. (lhs, rhs are just the column value)

- Create LongAnyAggregator, FloatAnyAggregator, DoubleAnyAggregator, StringAnyAggregator, LongAnyBufferAggregator, FloatAnyBufferAggregator, DoubleAnyBufferAggregator, StringAnyBufferAggregator. Aggregate logic will just return value if we already see non-null value without getObject from selector. LongAnyBufferAggregator, FloatAnyBufferAggregator, and DoubleAnyBufferAggregator have 0 as default values.

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+package org.apache.druid.query.aggregation.any;
+
+import org.apache.druid.query.aggregation.Aggregator;
+import org.apache.druid.segment.BaseDoubleColumnValueSelector;
+
+public class DoubleAnyAggregator implements Aggregator
+{
+  private final BaseDoubleColumnValueSelector valueSelector;
+
+  private Double foundValue;
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java]: I think you should store this as a primitive so we don't have to do un-necessary boxing. Also I think we need another variable in here to track whether the value is null or not. We would need to implement `isNull` for the aggregator.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  @Override
+  public Object get()
+  {
+    return foundValue;
+  }
+
+  @Override
+  public float getFloat()
+  {
+    return foundValue.floatValue();
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java]: This can throw an NPE if isNull is true. I can't tell if this is called from a hot loop or not. @clintropolis or @jon-wei might know how we should deal with this. I see this potential NPE in other aggregators as well.

If you switch to using a primitive in the class, then all these getters can simply cast the local variable

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  @Override
+  public double getDouble()
+  {
+    return foundValue;
+  }
+
+  @Override
+  public void close()
+  {
+
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java]: nit: Add a comment explaining why this function is empty to indicate it is intentional

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    DoubleAnyAggregatorFactory that = (DoubleAnyAggregatorFactory) o;
+
+    return name.equals(that.name) && fieldName.equals(that.fieldName);
+  }
+
+  @Override
+  public int hashCode()
+  {
+    return Objects.hash(name, fieldName);
+  }
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: Shouldn't we be looking at `storeDoubleAsFloat ` as well for equalsAndHashCode?

Can we add an EqualsVerifierTest for this?

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+  @Override
+  public String toString()
+  {
+    return "DoubleAnyAggregatorFactory{" +
+           "name='" + name + '\'' +
+           ", fieldName='" + fieldName + '\'' +
+           '}';
+  }
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: Similar comment to above...

Ugh I can't wait till we can hide all of this away with lombok

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    if (storeDoubleAsFloat) {
+      return "float";
+    }
+    return "double";
+  }
+
+  @Override
+  public int getMaxIntermediateSize()
+  {
+    return Double.BYTES;
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: not sure if this makes a difference or not, but won't the max size be `Float.BYTES` if storeDoubleAsFloat is true? 

I see the pattern you used is the same as what's used in `DoubleFirstAggregatorFactory` - probably a bug in both?

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+  @JsonProperty
+  public String getFieldName()
+  {
+    return fieldName;
+  }
+
+  @Override
+  public List<String> requiredFields()
+  {
+    return Arrays.asList(fieldName);
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: nit: `Collections.singletonList(fieldName)`

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+  @Override
+  public AggregatorFactory getCombiningFactory()
+  {
+    return new DoubleAnyAggregatorFactory(name, name);
+  }
+
+  @Override
+  public List<AggregatorFactory> getRequiredColumns()
+  {
+    return Collections.singletonList(new DoubleAnyAggregatorFactory(fieldName, fieldName));
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: I don't understand why both the parameters here are `fieldName` and both the parameters in getCombiningFactory are `name` - I see this pattern used in all the aggregators

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+  @Override
+  protected BufferAggregator factorizeBuffered(ColumnSelectorFactory metricFactory, ColumnValueSelector selector)
+  {
+    return new DoubleAnyBufferAggregator(selector);
+  }
+
+  @Override
+  public Comparator getComparator()
+  {
+    return DoubleAnyAggregatorFactory.VALUE_COMPARATOR;
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: Is this so that missing values show up first? Is that the behavior we want? I don't know the answer...

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+      @JsonProperty("name") String name,
+      @JsonProperty("fieldName") final String fieldName
+  )
+  {
+    Preconditions.checkNotNull(name, "Must have a valid, non-null aggregator name");
+    Preconditions.checkNotNull(fieldName, "Must have a valid, non-null fieldName");
+    this.name = name;
+    this.fieldName = fieldName;
+    this.storeDoubleAsFloat = ColumnHolder.storeDoubleAsFloat();
+  }
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: I know this is an existing pattern, but I prefer having a package private all args constructor to make unit testing easier and have the JsonCreator constructor call the all args constructor. Otherwise the tests need to rely on the static implementation of `ColumnHolder#storeDoubleAsFloat` which can be a huge pain to try and mock correctly.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyBufferAggregator.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  @Override
+  public double getDouble(ByteBuffer buf, int position)
+  {
+    return buf.getDouble(position);
+  }
+
+  @Override
+  public void close()
+  {
+
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyBufferAggregator.java]: nit: Comment for empty function please

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyBufferAggregator.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+import org.apache.druid.query.aggregation.BufferAggregator;
+import org.apache.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import org.apache.druid.segment.BaseDoubleColumnValueSelector;
+
+import java.nio.ByteBuffer;
+
+public class DoubleAnyBufferAggregator implements BufferAggregator
+{
+  private static final double NULL_VALUE = 0;
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyBufferAggregator.java]: 0 is a valid value for a double. I think `Double.NaN` is safer to indicate that the value is not found. 

Similar comments to `DoubleAnyAggregator` - I think you need to store a byte to indicate whether or not the value has been found since 0, NaN, etc. are all valid values that can show up in a double column.

File: processing/src/main/java/org/apache/druid/query/aggregation/first/StringAggregatorUtils.java [hunk 13]
```diff
@@ -87,7 +87,7 @@ public static void writePair(
 
     if (pair.rhs != null) {
       mutationBuffer.position(position + Long.BYTES + Integer.BYTES);
-      mutationBuffer.limit(maxStringBytes);
+      mutationBuffer.limit(position + Long.BYTES + Integer.BYTES + maxStringBytes);
```
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/StringAggregatorUtils.java]: Why was this change needed?

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java (hunk 2 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java]: (Double/Float/Long)AnyAggregators are created by the (Double/Float/Long)AnyAggregatorFactory. The (Double/Float/Long)AnyAggregatorFactory extends from NullableNumericAggregatorFactory. NullableNumericAggregatorFactory will handle the Null for the (Double/Float/Long)AnyAggregators. If useDefaultValueForNull=true then we don't have to worry about null since all null will be convert to non-null (i.e. 0) before aggregation. If useDefaultValueForNull=false then the NullableNumericAggregatorFactory will wrap the AnyAggregators in NullableNumericAggregator which already have isNull check. (Those methods like getFloat also won't be call if isNull is true)

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java (hunk 1 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java]: As mentioned before we don't have to worry about null. This is just considering which is more optimize between checking if Double is null or checking if boolean is T/F (and I guess memory for storing one Double vs. storing one double and one Boolean)

File: processing/src/main/java/org/apache/druid/query/aggregation/first/StringAggregatorUtils.java (hunk 13 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/StringAggregatorUtils.java]: This is to fix an existing bug i found. The limit method for buffer sets it at absolute position. The correct limit should be maxStringBytes after where we are writing the String to the buffer (which is position + Long.BYTES + Integer.BYTES). Hence, limit should be position + Long.BYTES + Integer.BYTES + maxStringBytes

File: processing/src/main/java/org/apache/druid/query/aggregation/AggregatorUtil.java [hunk 14]
```diff
@@ -121,6 +121,11 @@
 
   public static final byte MEAN_CACHE_TYPE_ID = 0x41;
 
+  public static final byte LONG_ANY_CACHE_TYPE_ID = 0x42;
+  public static final byte DOUBLE_ANY_CACHE_TYPE_ID = 0x42;
```
[the remark this answers is not in the document]
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/AggregatorUtil.java]: Copy+Paste and forgot to change the value. Done.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.aggregation.any;
+
+import org.apache.druid.query.aggregation.Aggregator;
+import org.apache.druid.segment.BaseDoubleColumnValueSelector;
+
+public class DoubleAnyAggregator implements Aggregator
```
[the remark this answers is not in the document]
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java]: Done.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java (hunk 1 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java]: Done.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java (hunk 3 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java]: Done

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java (hunk 4 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: No idea. DoubleFirstAggregatorFactory does not look at storeDoubleAsFloat for hashCode.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java (hunk 5 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: No idea. DoubleFirstAggregatorFactory does not look at storeDoubleAsFloat for hashCode.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java (hunk 7 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: Done

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java (hunk 8 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: There's javadoc on AggregatorFactory class. But i think it's like reusing input output fields

Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: getCombiningFactory are combining between segments

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java (hunk 9 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: Not sure I understand the question. This is for sorting the aggregated result. I think it does not really matters

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java (hunk 10 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: Not sure if the storeDoubleAsFloat is even needed.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyBufferAggregator.java (hunk 11 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyBufferAggregator.java]: Done

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java (hunk 4 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: Actually gonna remove storeDoubleAsFloat

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java (hunk 5 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: Actually gonna remove storeDoubleAsFloat

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java (hunk 6 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: Actually gonna remove storeDoubleAsFloat

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java (hunk 10 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: Actually gonna remove storeDoubleAsFloat

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyBufferAggregator.java (hunk 12 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyBufferAggregator.java]: Added byte to indicate whether or not the value has been found. We actually do not get null due to the NullableNumericAggregatorFactory/ NullableNumericAggregator stuff

File: processing/src/main/java/org/apache/druid/query/aggregation/first/StringAggregatorUtils.java (hunk 13 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/StringAggregatorUtils.java]: Removed this. Will fix as part of separate PR

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+  public DoubleAnyAggregator(BaseDoubleColumnValueSelector valueSelector)
+  {
+    this.valueSelector = valueSelector;
+    foundValue = null;
+  }
+
+  @Override
+  public void aggregate()
+  {
+    if (foundValue == null && !valueSelector.isNull()) {
```
[the remark this answers is not in the document]
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregator.java]: Removed.

**Comment by maytasm3** [comment]
Note: ANY on Double/Float/Long column with the useDefaultValueForNull=true will **not** prefer non-null values over default value for null (i.e. 0)

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+
+public class DoubleAnyAggregatorFactory extends NullableNumericAggregatorFactory<ColumnValueSelector>
```
[the remark this answers is not in the document]
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: Done

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+  @Override
+  public List<AggregatorFactory> getRequiredColumns()
+  {
+    return Collections.singletonList(new DoubleAnyAggregatorFactory(fieldName, fieldName, expression, macroTable));
+  }
+
+  @Override
+  public byte[] getCacheKey()
+  {
+    byte[] fieldNameBytes = StringUtils.toUtf8WithNullToEmpty(fieldName);
```
Comment (jon-wei) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: This method and similar could use `CacheKeyBuilder` instead

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyBufferAggregator.java [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    buf.put(position, BYTE_FLAG_IS_NOT_SET);
+    buf.putDouble(position + Byte.BYTES, NULL_VALUE);
+  }
+
+  @Override
+  public void aggregate(ByteBuffer buf, int position)
+  {
+    if (buf.get(position) == BYTE_FLAG_IS_NOT_SET) {
+      buf.putDouble(position + Byte.BYTES, valueSelector.getDouble());
+      buf.put(position, BYTE_FLAG_IS_SET);
```
Comment (jon-wei) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyBufferAggregator.java]: nit: suggest setting the IS_SET byte first before the double value, since that's the order they appear in the buffer

File: docs/querying/sql.md [hunk 20]
```diff
@@ -203,6 +203,10 @@ Only the COUNT aggregation can accept DISTINCT.
...
 |`LATEST(expr)`|Returns the latest non-null value of `expr`, which must be numeric. If `expr` comes from a relation with a timestamp column (like a Druid datasource) then "latest" is the value last encountered with the maximum overall timestamp of all values being aggregated. If `expr` does not come from a relation with a timestamp, then it is simply the last value encountered.|
 |`LATEST(expr, maxBytesPerString)`|Like `LATEST(expr)`, but for strings. The `maxBytesPerString` parameter determines how much aggregation space to allocate per string. Strings longer than this limit will be truncated. This parameter should be set as low as possible, since high values will lead to wasted memory.|
+|`ANY_VALUE(expr)`|Returns any value of `expr`, which must be numeric. If `druid.generic.useDefaultValueForNull=true` this can returns the default value for null and does not prefer "non-null" values over the default value for null. If `druid.generic.useDefaultValueForNull=false`, then this will returns any non-null value of `expr`|
```
Comment (jon-wei) [inline docs/querying/sql.md]: "this can returns the default" -> "this can return the default", similarly for "then this will returns"

File: docs/querying/sql.md [hunk 21]
```diff
@@ -203,6 +203,10 @@ Only the COUNT aggregation can accept DISTINCT.
...
+|`ANY_VALUE(expr, maxBytesPerString)`|Like `ANY_VALUE(expr)`, but for strings. The `maxBytesPerString` parameter determines how much aggregation space to allocate per string. Strings longer than this limit will be truncated. This parameter should be set as low as possible, since high values will lead to wasted memory.|
```
Comment (jon-wei) [inline docs/querying/sql.md]: This should mention that the default maxBytesPerString is 1024

File: docs/querying/sql.md (hunk 20 above)
Comment (jon-wei) [inline docs/querying/sql.md]: Can you also add entries for the new aggregators under `docs/querying/aggregations.md`?

File: processing/src/main/java/org/apache/druid/query/aggregation/any/StringAnyAggregator.java [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+          foundValue = foundValue.substring(0, maxStringBytes);
+        }
+      }
+    }
+  }
+
+  @Override
+  public Object get()
+  {
+    return StringAggregatorUtils.chop(foundValue, maxStringBytes);
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/StringAnyAggregator.java]: nit: It is probably worth pushing `chop` down into `StringUtils` rather than renaming and widening the usage of `StringAggregatorUtils`

File: processing/src/main/java/org/apache/druid/query/aggregation/first/StringAggregatorUtils.java [hunk 23]
```diff
@@ -29,7 +29,7 @@
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
 
-public class StringFirstLastUtils
+public class StringAggregatorUtils
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/StringAggregatorUtils.java]: If you decide to end up moving `chop` to `StringUtils`, please revert this rename

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java [hunk 24]
```diff
@@ -1297,6 +1301,46 @@ public void testLatestAggregators() throws Exception
     );
   }
 
+  // This test the on-heap version of the AnyAggregator (Double/Float/Long/String)
+  @Test
+  public void testAnyAggregator() throws Exception
```
Comment (clintropolis) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java]: nit: It would probably be worth adding an additional test that tests numeric columns agains `druid.numfoo` table since it contains numeric columns that have null values when run in sql compatible null mode, and also tests for ordering by each 'any' aggregator.

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java (hunk 18 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyAggregatorFactory.java]: Done!

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyBufferAggregator.java (hunk 19 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyBufferAggregator.java]: Done

File: docs/querying/sql.md (hunk 20 above)
Comment (maytasm3) [reply docs/querying/sql.md]: Done

File: docs/querying/sql.md (hunk 21 above)
Comment (maytasm3) [reply docs/querying/sql.md]: What do you mean default? You need to always pass the value. There is no default value for maxBytesPerString

File: docs/querying/sql.md (hunk 20 above)
Comment (maytasm3) [reply docs/querying/sql.md]: Done. Btw I saw filterNullValues for stringLast and stringFirst. Is that still true?

File: processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyBufferAggregator.java [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+  public DoubleAnyBufferAggregator(BaseDoubleColumnValueSelector valueSelector)
+  {
+    this.valueSelector = valueSelector;
+  }
+
+  @Override
+  public void init(ByteBuffer buf, int position)
+  {
+    buf.put(position, BYTE_FLAG_IS_NOT_SET);
+    buf.putDouble(position + Byte.BYTES, NULL_VALUE);
```
[the remark this answers is not in the document]
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/DoubleAnyBufferAggregator.java]: Done

File: processing/src/main/java/org/apache/druid/query/aggregation/any/StringAnyAggregator.java (hunk 22 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/StringAnyAggregator.java]: sounds good to me. Done

File: processing/src/main/java/org/apache/druid/query/aggregation/first/StringAggregatorUtils.java (hunk 23 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/StringAggregatorUtils.java]: done

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java (hunk 24 above)
Comment (maytasm3) [reply sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java]: done

File: processing/src/main/java/org/apache/druid/query/aggregation/any/StringAnyAggregator.java [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+    this.maxStringBytes = maxStringBytes;
+    this.foundValue = null;
+  }
+
+  @Override
+  public void aggregate()
+  {
+    if (foundValue == null) {
+      final Object object = valueSelector.getObject();
+      if (object != null) {
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/StringAnyAggregator.java]: sorry I missed this earlier, this check isn't necessary, `DimensionHandlerUtils.convertObjectToString` has it's own null check

File: processing/src/main/java/org/apache/druid/query/aggregation/any/StringAnyBufferAggregator.java [hunk 27]
```diff
@@ -0,0 +1,104 @@
...
+    buf.putInt(position, NULL_STRING_LENGTH);
+  }
+
+  @Override
+  public void aggregate(ByteBuffer buf, int position)
+  {
+    int stringSizeBytes = buf.getInt(position);
+    if (stringSizeBytes < 0) {
+      final Object object = valueSelector.getObject();
+      if (object != null) {
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/aggregation/any/StringAnyBufferAggregator.java]: same comment about unnecessary check

File: sql/src/test/java/org/apache/druid/sql/calcite/util/CalciteTests.java [hunk 28]
```diff
@@ -377,6 +377,15 @@ public AuthenticationResult createEscalatedAuthenticationResult()
   );
 
   public static final List<InputRow> ROWS1_WITH_NUMERIC_DIMS = ImmutableList.of(
+      createRow(
```
Comment (clintropolis) [inline sql/src/test/java/org/apache/druid/sql/calcite/util/CalciteTests.java]: Hmm, it seems like this change is causing some unrelated test failures

File: processing/src/main/java/org/apache/druid/query/aggregation/any/StringAnyAggregator.java (hunk 26 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/StringAnyAggregator.java]: Done

File: processing/src/main/java/org/apache/druid/query/aggregation/any/StringAnyBufferAggregator.java (hunk 27 above)
Comment (maytasm3) [reply processing/src/main/java/org/apache/druid/query/aggregation/any/StringAnyBufferAggregator.java]: Done

File: sql/src/test/java/org/apache/druid/sql/calcite/util/CalciteTests.java (hunk 28 above)
Comment (maytasm3) [reply sql/src/test/java/org/apache/druid/sql/calcite/util/CalciteTests.java]: Seems like the VarianceSqlAggregatorTest is using this data too and when the ordering of the rows changed, the variance also changed. I created a new datasource that have the numeric dim first for my test and changed the numfoo datasource back to how it was. The reason I wanted to have numeric null first is because the ANY will select the first row and skip everything after. So if the first row is not null, then there is not really any point in testing (if we want to test the numeric null stuff)

File: docs/querying/sql.md (hunk 21 above)
Comment (jon-wei) [reply docs/querying/sql.md]: you have this block in StringAnyAggregatorFactory:

```

    this.maxStringBytes = maxStringBytes == null
                          ? StringFirstAggregatorFactory.DEFAULT_MAX_STRING_SIZE
                          : maxStringBytes;
```

I would give the SQL function consistent behavior

Comment (maytasm3) [reply docs/querying/sql.md]: Currently, the implementation for LATEST, EARLIEST (and ANY since I based it off LATEST, EARLIEST) is that if you use the json stuff, then maxStringBytes is optional and if not present will default to 1024 (as per the docs in docs/querying/aggregations.md). 
However, this does not work the same if you issue the query through SQL. To use LATEST, EARLIEST (and ANY) in SQL, you must give the maxStringBytes as the second argument. If you do not, then the column actually gets cast into double (super weird).

Comment (maytasm3) [reply docs/querying/sql.md]: Let's discuss. We can change this behaviour for LATEST, EARLIEST (and ANY)

File: sql/src/test/java/org/apache/druid/sql/calcite/util/CalciteTests.java (hunk 28 above)
Comment (maytasm3) [reply sql/src/test/java/org/apache/druid/sql/calcite/util/CalciteTests.java]: Actually, I think it's fine to just test with the same numfoo datasource (with first row being non-null)

File: docs/querying/sql.md (hunk 20 above)
Comment (jon-wei) [reply docs/querying/sql.md]: Hm, looks like the docs are out of date for those, we can fix those later

**Review by EnsDeLiz — COMMENTED** [review]
@jihoonson
